File: packages/browser/src/browser-use-backend.ts
import { spawn } from "node:child_process";

import { BrowserActionError, BrowserNotAvailable } from "./errors.js";
import type { BrowserBackend, BrowserElement, BrowserState } from "./types.js";

const COMMAND_TIMEOUT_MS = 30_000;
const ELEMENT_LINE_RE = /^\s*\[(\d+)]\s+([^\s]+)(?:\s+"([^"]*)")?/;
const URL_LINE_RE = /^\s*(?:url|current url|page url):\s*(\S+)/i;

interface CommandResult {
  stdout: string;
  stderr: string;
}

function runBrowserUse(args: string[]): Promise<CommandResult> {
  return new Promise((resolve, reject) => {
    const child = spawn("browser-use", args, { stdio: ["ignore", "pipe", "pipe"] });
    const stdoutChunks: Buffer[] = [];
    const stderrChunks: Buffer[] = [];

    const timeout = setTimeout(() => {
      child.kill("SIGTERM");
      reject(new BrowserActionError(`browser-use ${args.join(" ")} timed out after 30s`));
    }, COMMAND_TIMEOUT_MS);

    child.stdout.on("data", (chunk: Buffer) => stdoutChunks.push(chunk));
    child.stderr.on("data", (chunk: Buffer) => stderrChunks.push(chunk));
    child.on("error", (error: NodeJS.ErrnoException) => {
      clearTimeout(timeout);
      if (error.code === "ENOENT") {
        reject(new BrowserNotAvailable("browser-use CLI is not available on PATH"));
        return;
      }
      reject(new BrowserActionError(error.message));
    });
    child.on("close", (code) => {
      clearTimeout(timeout);
      const stdout = Buffer.concat(stdoutChunks).toString("utf-8");
      const stderr = Buffer.concat(stderrChunks).toString("utf-8");
      if (code !== 0) {
        reject(new BrowserActionError((stderr || stdout || `browser-use exited with code ${code}`).trim()));
        return;
      }
      resolve({ stdout, stderr });
    });
  });
}

function parseState(stdout: string): BrowserState {
  const elements: BrowserElement[] = [];
  let url = "";

  for (const line of stdout.split("\n")) {
    const urlMatch = URL_LINE_RE.exec(line);
    if (urlMatch?.[1] !== undefined) {
      url = urlMatch[1];
      continue;
    }

    const match = ELEMENT_LINE_RE.exec(line);
    if (match?.[1] === undefined || match[2] === undefined) {
      continue;
    }
    elements.push({ id: match[1], role: match[2], text: match[3] });
  }

  return { url, elements };
}

function parseEval(stdout: string): unknown {
  const trimmed = stdout.trim();
  if (trimmed.length === 0) {
    return undefined;
  }
  try {
    return JSON.parse(trimmed);
  } catch {
    return trimmed;
  }
}

export class BrowserUseBackend implements BrowserBackend {
  async open(url: string): Promise<void> {
    await runBrowserUse(["open", url]);
  }

  async getState(): Promise<BrowserState> {
    const result = await runBrowserUse(["state"]);
    return parseState(result.stdout);
  }

  async click(elementId: string): Promise<void> {
    await runBrowserUse(["click", elementId]);
  }

  async type(elementId: string, text: string): Promise<void> {
    await runBrowserUse(["input", elementId, text]);
  }

  async evaluate(script: string): Promise<unknown> {
    const result = await runBrowserUse(["eval", script]);
    return parseEval(result.stdout);
  }

  async close(): Promise<void> {
    await runBrowserUse(["close"]);
  }
}

File: packages/browser/browser/browser_use_backend.py
from __future__ import annotations

import json
import re
import subprocess
from typing import Any

from browser.errors import BrowserActionError, BrowserNotAvailable
from browser.types import BrowserBackend, BrowserElement, BrowserState

COMMAND_TIMEOUT_SECONDS = 30
ELEMENT_LINE_RE = re.compile(r'^\s*\[(\d+)]\s+([^\s]+)(?:\s+"([^"]*)")?')
URL_LINE_RE = re.compile(r"^\s*(?:url|current url|page url):\s*(\S+)", re.IGNORECASE)


def _run_browser_use(args: list[str]) -> subprocess.CompletedProcess[str]:
    command = ["browser-use", *args]
    try:
        result = subprocess.run(command, capture_output=True, check=False, text=True, timeout=COMMAND_TIMEOUT_SECONDS)
    except FileNotFoundError as exc:
        raise BrowserNotAvailable("browser-use CLI is not available on PATH") from exc
    except subprocess.TimeoutExpired as exc:
        raise BrowserActionError(f"browser-use {' '.join(args)} timed out after 30s") from exc

    if result.returncode != 0:
        output = result.stderr or result.stdout or f"browser-use exited with code {result.returncode}"
        raise BrowserActionError(output.strip())
    return result


def _parse_state(stdout: str) -> BrowserState:
    elements: list[BrowserElement] = []
    url = ""
    for line in stdout.splitlines():
        url_match = URL_LINE_RE.match(line)
        if url_match:
            url = url_match.group(1)
            continue

        match = ELEMENT_LINE_RE.match(line)
        if not match:
            continue
        elements.append(BrowserElement(id=match.group(1), role=match.group(2), text=match.group(3)))
    return BrowserState(url=url, elements=elements)


def _parse_eval(stdout: str) -> Any:
    trimmed = stdout.strip()
    if not trimmed:
        return None
    try:
        return json.loads(trimmed)
    except json.JSONDecodeError:
        return trimmed


class BrowserUseBackend(BrowserBackend):
    def open(self, url: str) -> None:
        _run_browser_use(["open", "--", url])

    def get_state(self) -> BrowserState:
        result = _run_browser_use(["state"])
        try:
            return _parse_state(result.stdout)
        except (ValueError, TypeError, AttributeError) as exc:
            raise BrowserActionError(f"Failed to parse browser-use state: {exc}") from exc

    def click(self, element_id: str) -> None:
        _run_browser_use(["click", "--", element_id])

    def type(self, element_id: str, text: str) -> None:
        _run_browser_use(["input", "--", element_id, text])

    def evaluate(self, script: str) -> Any:
        result = _run_browser_use(["eval", "--", script])
        return _parse_eval(result.stdout)

    def close(self) -> None:
        _run_browser_use(["close"])

    def wait_for_selector(self, selector: str, state: str = "visible", timeout: int = 5000) -> bool:
        raise BrowserActionError("wait_for_selector not supported by browser-use")

    def capture_screenshot(self, path: str | None = None) -> bytes:
        raise BrowserActionError("capture_screenshot not supported by browser-use")

    def save_storage_state(self, path: str) -> None:
        raise BrowserActionError("save_storage_state not supported by browser-use")

    def scroll_to(self, element_id: str) -> None:
        raise BrowserActionError("scroll_to not supported by browser-use")
